File: tests/domains/video/test_metadata.py
import json
import subprocess

import pytest

from domains import Runtime
from domains.video.metadata import set_video_metadata
from tests.helpers import unwrap

pytestmark = pytest.mark.ffmpeg


def test_set_metadata(runtime: Runtime, sample_video: str) -> None:
    result = unwrap(
        set_video_metadata(
            runtime, sample_video, title="Meu título", description="Descrição longa", author="Eu"
        )
    )
    assert result["output"] == "sample_meta.mp4"
    raw = subprocess.run(
        [
            "ffprobe",
            "-v",
            "error",
            "-print_format",
            "json",
            "-show_format",
            str(runtime.workspace.existing(result["output"])),
        ],
        capture_output=True,
        text=True,
        check=True,
    ).stdout
    tags = json.loads(raw)["format"]["tags"]
    assert tags["title"] == "Meu título"
    assert tags["description"] == "Descrição longa"
    assert tags["artist"] == "Eu"


def test_set_metadata_nothing(runtime: Runtime, sample_video: str) -> None:
    assert set_video_metadata(runtime, sample_video).get("code") == "invalid_argument"


def test_set_metadata_too_long(runtime: Runtime, sample_video: str) -> None:
    result = set_video_metadata(runtime, sample_video, description="x" * 6000)
    assert result.get("code") == "invalid_argument"


def test_set_metadata_missing(runtime: Runtime) -> None:
    assert set_video_metadata(runtime, "ghost.mp4", title="x").get("code") == "not_found"

File: tests/domains/video/test_metadata.rs
use std::process::Command;

use mcp_tools::core::errors::ErrorCode;
use mcp_tools::domains::video::metadata::{set_video_metadata, MetadataFields};

use crate::conftest::{runtime, sample_video};
use crate::skip_without_ffmpeg;

#[test]
fn test_set_metadata() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let fields = MetadataFields {
        title: Some("Meu título".to_string()),
        description: Some("Descrição longa".to_string()),
        author: Some("Eu".to_string()),
        comment: None,
    };
    let result = set_video_metadata(&rt.runtime, &sample, &fields).unwrap();
    assert_eq!(result.output, "sample_meta.mp4");
    let target = rt.runtime.workspace.existing(&result.output).unwrap();
    let output = Command::new("ffprobe")
        .args(["-v", "error", "-print_format", "json", "-show_format"])
        .arg(&target)
        .output()
        .expect("ffprobe");
    assert!(output.status.success());
    let raw: serde_json::Value = serde_json::from_slice(&output.stdout).unwrap();
    let tags = &raw["format"]["tags"];
    assert_eq!(tags["title"], "Meu título");
    assert_eq!(tags["description"], "Descrição longa");
    assert_eq!(tags["artist"], "Eu");
}

#[test]
fn test_set_metadata_nothing() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let error = set_video_metadata(&rt.runtime, &sample, &MetadataFields::default()).unwrap_err();
    assert_eq!(error.code, ErrorCode::InvalidArgument);
}

#[test]
fn test_set_metadata_too_long() {
    skip_without_ffmpeg!();
    let rt = runtime();
    let sample = sample_video(&rt.runtime.workspace);
    let fields = MetadataFields {
        description: Some("x".repeat(6000)),
        ..MetadataFields::default()
    };
    let error = set_video_metadata(&rt.runtime, &sample, &fields).unwrap_err();
    assert_eq!(error.code, ErrorCode::InvalidArgument);
}

#[test]
fn test_set_metadata_missing() {
    let rt = runtime();
    let fields = MetadataFields {
        title: Some("x".to_string()),
        ..MetadataFields::default()
    };
    let error = set_video_metadata(&rt.runtime, "ghost.mp4", &fields).unwrap_err();
    assert_eq!(error.code, ErrorCode::NotFound);
}
